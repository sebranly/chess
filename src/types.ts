export enum PieceType {
  Bishop = 'bishop',
  King = 'king',
  Knight = 'knight',
  Pawn = 'pawn',
  Queen = 'queen',
  Rook = 'rook',
}
