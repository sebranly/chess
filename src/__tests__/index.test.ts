import {
  getCell,
  getFileBasedPieceType,
  getFileBasedPieceSubType,
  getFileLetter,
  getPiecePoints,
  getTerminalNotation,
  getTerminalNotationLetter,
  initializeBoard,
  initializePiece,
} from '../index';
import { Cell, Color, Piece, PieceSubType, PieceType } from '../types';

const { Bishop, King, Knight, Pawn, Queen, Rook } = PieceType;

test('getPiecePoints', () => {
  expect(getPiecePoints(King)).toBe(0);
  expect(getPiecePoints(Pawn)).toBe(1);
  expect(getPiecePoints(Bishop)).toBe(3);
  expect(getPiecePoints(Knight)).toBe(3);
  expect(getPiecePoints(Rook)).toBe(5);
  expect(getPiecePoints(Queen)).toBe(9);
});

test('getFileLetter', () => {
  expect(getFileLetter(0)).toBe('');
  expect(getFileLetter(1)).toBe('a');
  expect(getFileLetter(8)).toBe('h');
  expect(getFileLetter(26)).toBe('z');
});

test('initializePiece', () => {
  const whitePawn = initializePiece(Color.White);
  const expectedWhitePawn: Piece = {
    color: Color.White,
    hasNeverMoved: true,
    isFromPromotion: false,
    subType: undefined,
    type: PieceType.Pawn,
  };

  expect(whitePawn).toStrictEqual(expectedWhitePawn);

  const eastBlackRook = initializePiece(Color.Black, PieceType.Rook, PieceSubType.East);
  const expectedEastBlackRook: Piece = {
    color: Color.Black,
    hasNeverMoved: true,
    isFromPromotion: false,
    subType: PieceSubType.East,
    type: PieceType.Rook,
  };

  expect(eastBlackRook).toStrictEqual(expectedEastBlackRook);
});

test('getFileBasedPieceType', () => {
  expect(getFileBasedPieceType(1)).toBe(Rook);
  expect(getFileBasedPieceType(2)).toBe(Knight);
  expect(getFileBasedPieceType(3)).toBe(Bishop);
  expect(getFileBasedPieceType(4)).toBe(Queen);
  expect(getFileBasedPieceType(5)).toBe(King);
  expect(getFileBasedPieceType(6)).toBe(Bishop);
  expect(getFileBasedPieceType(7)).toBe(Knight);
  expect(getFileBasedPieceType(8)).toBe(Rook);
});

test('getFileBasedPieceSubType', () => {
  const { West, East } = PieceSubType;

  expect(getFileBasedPieceSubType(1)).toBe(West);
  expect(getFileBasedPieceSubType(2)).toBe(West);
  expect(getFileBasedPieceSubType(3)).toBe(West);
  expect(getFileBasedPieceSubType(4)).toBeUndefined();
  expect(getFileBasedPieceSubType(5)).toBeUndefined();
  expect(getFileBasedPieceSubType(6)).toBe(East);
  expect(getFileBasedPieceSubType(7)).toBe(East);
  expect(getFileBasedPieceSubType(8)).toBe(East);
});

test('initializeBoard', () => {
  expect(initializeBoard()).toMatchSnapshot();
});

test('getCell', () => {
  const board = initializeBoard();

  const expectedPiece: Piece = {
    color: Color.White,
    hasNeverMoved: true,
    isFromPromotion: false,
    subType: PieceSubType.West,
    type: PieceType.Rook,
  };

  const expectedCell: Cell = {
    file: 'a',
    rank: 1,
    piece: expectedPiece,
  };

  expect(getCell(board, 'a', 1)).toStrictEqual(expectedCell);
  expect(getCell(board, 'z', 1)).toBeUndefined();
});

test('getTerminalNotationLetter', () => {
  expect(getTerminalNotationLetter(King)).toBe('k');
  expect(getTerminalNotationLetter(Pawn)).toBe('p');
  expect(getTerminalNotationLetter(Bishop)).toBe('b');
  expect(getTerminalNotationLetter(Knight)).toBe('n');
  expect(getTerminalNotationLetter(Rook)).toBe('r');
  expect(getTerminalNotationLetter(Queen)).toBe('q');
});

test('getTerminalNotation', () => {
  const piece: Piece = {
    color: Color.White,
    hasNeverMoved: true,
    isFromPromotion: false,
    subType: undefined,
    type: PieceType.King,
  };

  expect(getTerminalNotation(piece)).toBe('K');

  piece.color = Color.Black;
  expect(getTerminalNotation(piece)).toBe('k');
});
