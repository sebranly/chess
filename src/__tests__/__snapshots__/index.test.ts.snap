// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializeBoard 1`] = `
Object {
  "cells": Array [
    Object {
      "file": "a",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "rook",
      },
      "rank": 1,
    },
    Object {
      "file": "a",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "a",
      "rank": 3,
    },
    Object {
      "file": "a",
      "rank": 4,
    },
    Object {
      "file": "a",
      "rank": 5,
    },
    Object {
      "file": "a",
      "rank": 6,
    },
    Object {
      "file": "a",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "a",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "rook",
      },
      "rank": 8,
    },
    Object {
      "file": "b",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "knight",
      },
      "rank": 1,
    },
    Object {
      "file": "b",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "b",
      "rank": 3,
    },
    Object {
      "file": "b",
      "rank": 4,
    },
    Object {
      "file": "b",
      "rank": 5,
    },
    Object {
      "file": "b",
      "rank": 6,
    },
    Object {
      "file": "b",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "b",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "knight",
      },
      "rank": 8,
    },
    Object {
      "file": "c",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "bishop",
      },
      "rank": 1,
    },
    Object {
      "file": "c",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "c",
      "rank": 3,
    },
    Object {
      "file": "c",
      "rank": 4,
    },
    Object {
      "file": "c",
      "rank": 5,
    },
    Object {
      "file": "c",
      "rank": 6,
    },
    Object {
      "file": "c",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "c",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "west",
        "type": "bishop",
      },
      "rank": 8,
    },
    Object {
      "file": "d",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "queen",
      },
      "rank": 1,
    },
    Object {
      "file": "d",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "d",
      "rank": 3,
    },
    Object {
      "file": "d",
      "rank": 4,
    },
    Object {
      "file": "d",
      "rank": 5,
    },
    Object {
      "file": "d",
      "rank": 6,
    },
    Object {
      "file": "d",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "d",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "queen",
      },
      "rank": 8,
    },
    Object {
      "file": "e",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "king",
      },
      "rank": 1,
    },
    Object {
      "file": "e",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "e",
      "rank": 3,
    },
    Object {
      "file": "e",
      "rank": 4,
    },
    Object {
      "file": "e",
      "rank": 5,
    },
    Object {
      "file": "e",
      "rank": 6,
    },
    Object {
      "file": "e",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "e",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "king",
      },
      "rank": 8,
    },
    Object {
      "file": "f",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "bishop",
      },
      "rank": 1,
    },
    Object {
      "file": "f",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "f",
      "rank": 3,
    },
    Object {
      "file": "f",
      "rank": 4,
    },
    Object {
      "file": "f",
      "rank": 5,
    },
    Object {
      "file": "f",
      "rank": 6,
    },
    Object {
      "file": "f",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "f",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "bishop",
      },
      "rank": 8,
    },
    Object {
      "file": "g",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "knight",
      },
      "rank": 1,
    },
    Object {
      "file": "g",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "g",
      "rank": 3,
    },
    Object {
      "file": "g",
      "rank": 4,
    },
    Object {
      "file": "g",
      "rank": 5,
    },
    Object {
      "file": "g",
      "rank": 6,
    },
    Object {
      "file": "g",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "g",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "knight",
      },
      "rank": 8,
    },
    Object {
      "file": "h",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "rook",
      },
      "rank": 1,
    },
    Object {
      "file": "h",
      "piece": Object {
        "color": "white",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 2,
    },
    Object {
      "file": "h",
      "rank": 3,
    },
    Object {
      "file": "h",
      "rank": 4,
    },
    Object {
      "file": "h",
      "rank": 5,
    },
    Object {
      "file": "h",
      "rank": 6,
    },
    Object {
      "file": "h",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": undefined,
        "type": "pawn",
      },
      "rank": 7,
    },
    Object {
      "file": "h",
      "piece": Object {
        "color": "black",
        "hasNeverMoved": true,
        "isFromPromotion": false,
        "subType": "east",
        "type": "rook",
      },
      "rank": 8,
    },
  ],
  "fileCount": 8,
  "rankCount": 8,
}
`;
